django_comments/views/comments.py

```diff
@@ -27,105 +33,157 @@ def __init__(self, why):
-    if form.errors or preview:
-        template_list = [
-            # These first two exist for purely historical reasons.
-            # Django v1.0 and v1.1 allowed the underscore format for
-            # preview templates, so we have to preserve that format.
-            "comments/%s_%s_preview.html" % (model._meta.app_label, model._meta.model_name),
-            "comments/%s_preview.html" % model._meta.app_label,
-            # Now the usual directory based template hierarchy.
-            "comments/%s/%s/preview.html" % (model._meta.app_label, model._meta.model_name),
-            "comments/%s/preview.html" % model._meta.app_label,
-            "comments/preview.html",
-        ]
-        return render(request, template_list, {
-                "comment": form.data.get("comment", ""),
-                "form": form,
-                "next": data.get("next", next),
-            },
+    def __init__(self, why):
+        self.response = CommentPostBadRequest(why)
+
+
+class CommentPostView(FormView):
+    http_method_names = ['post']
+
+    def get_target_object(self, data):
+        # Look up the object we're trying to comment about
+        ctype = data.get("content_type")
+        object_pk = data.get("object_pk")
+        if ctype is None or object_pk is None:
+            raise BadRequest("Missing content_type or object_pk field.")
+        try:
+            model = apps.get_model(*ctype.split(".", 1))
+            return model._default_manager.using(self.kwargs.get('using')).get(pk=object_pk)
+        except TypeError:
+            raise BadRequest("Invalid content_type value: %r" % escape(ctype))
+        except AttributeError:
+            raise BadRequest("The given content-type %r does not resolve to a valid model." % escape(ctype))
+        except ObjectDoesNotExist:
+            raise BadRequest("No object matching content-type %r and object PK %r exists." % (
+                    escape(ctype), escape(object_pk)))
+        except (ValueError, ValidationError) as e:
+            raise BadRequest("Attempting to get content-type %r and object PK %r raised %s" % (
+                    escape(ctype), escape(object_pk), e.__class__.__name__))
+
+    def get_form_kwargs(self):
+        data = self.request.POST.copy()
+        if self.request.user.is_authenticated:
+            if not data.get('name', ''):
+                data["name"] = self.request.user.get_full_name() or self.request.user.get_username()
+            if not data.get('email', ''):
+                data["email"] = self.request.user.email
+        return data
+
+    def get_form_class(self):
+        """Return the form class to use."""
+        return django_comments.get_form()
+
+    def get_form(self, form_class=None):
+        """Return an instance of the form to be used in this view."""
+        if form_class is None:
+            form_class = self.get_form_class()
+        return form_class(self.target_object, data=self.data)
```

Comment:
Can this invocation be modeled after the inherited `get_form()` implementation and move self.target_object and self.data as part of get_form_kwargs ? The parent implementation is
```
    def get_form(self, form_class=None):
        """Return an instance of the form to be used in this view."""
        if form_class is None:
            form_class = self.get_form_class()
        return form_class(**self.get_form_kwargs())
```

Reply:
Since we need the data coming from self.get_form_kwargs() and the target-object in various places in the code I decided to assign them to self.data and self.target_object in the post-method.

In my opinion this is the most comprehensible solution.

Reply:
Another thing is, that the comment form class takes the target_object as positional argument, which cannot be passed by unpacking the dict returned by self.get_form_kwargs().